File: src/features/typing/data/go/binary_tree_maximum_path_sum.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	best := root.Val
	var gain func(node *TreeNode) int
	gain = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		l := max(gain(node.Left), 0)
		r := max(gain(node.Right), 0)
		if node.Val+l+r > best {
			best = node.Val + l + r
		}
		return node.Val + max(l, r)
	}
	gain(root)
	return best
}

File: src/features/typing/data/ts/binary_tree_maximum_path_sum.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function maxPathSum(root: TreeNode | null): number {
    let best = -Infinity;
    const gain = (node: TreeNode | null): number => {
        if (node === null) {
            return 0;
        }
        const l = Math.max(gain(node.left), 0);
        const r = Math.max(gain(node.right), 0);
        if (node.val + l + r > best) {
            best = node.val + l + r;
        }
        return node.val + Math.max(l, r);
    };
    gain(root);
    return best;
}
